ttnn/cpp/ttnn/operations/experimental/reshape/reshape.cpp

```diff
@@ -0,0 +1,163 @@
+
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttnn/common/constants.hpp"
+#include "ttnn/run_operation.hpp"
+#include "reshape.hpp"
+#include "tt_metal/common/constants.hpp"
+#include <functional>
+#include <ttnn/operations/numpy/functions.hpp>
+#include "ttnn/operations/experimental/auto_format/auto_format.hpp"
+#include "ttnn/tensor/tensor_utils.hpp"
+#include "ttnn/operations/data_movement/data_transfer/data_transfer.hpp"
+#include "ttnn/operations/data_movement/slice/slice.hpp"
+#include "ttnn/operations/core/core.hpp"
+
+
+#include "ttnn/tensor/tensor.hpp"
+
+#include <cstdint>
+#include <memory>
+
+#include "common/bfloat16.hpp"
+#include "ttnn/tensor/tensor_impl.hpp"
+#include "ttnn/tensor/tensor_impl_wrapper.hpp"
+#include "ttnn/tensor/tensor_utils.hpp"
+#include "ttnn/tensor/types.hpp"
+#include "tt_metal/common/constants.hpp"
+#include "tt_metal/common/math.hpp"
+#include "tt_metal/third_party/tracy/public/tracy/Tracy.hpp"
+#include "tt_metal/graph/graph_tracking.hpp"
+#include "ttnn/distributed/api.hpp"
+#include "ttnn/distributed/types.hpp"
+#include "ttnn/core.hpp"
+
+namespace ttnn::operations::experimental::reshape {
+
+ttnn::Tensor MultiDeviceHostStorageVisit(
+    auto&& storage, const ttnn::Tensor& input_tensor, const ttnn::Shape& new_shape) {
+    using T = std::decay_t<decltype(storage)>;
+    auto updated_storage = std::get<T>(input_tensor.get_storage());
+    for (int i = 0; i < updated_storage.shapes.size(); i++) {
+        updated_storage.shapes[i] = new_shape;
+    }
+    if (input_tensor.get_layout() == Layout::ROW_MAJOR) {
+        return Tensor(updated_storage, new_shape, input_tensor.get_dtype(), input_tensor.get_layout(), std::nullopt);
+    } else {
+        const auto tile = input_tensor.get_tensor_spec().tile();
+        return Tensor(updated_storage, new_shape, input_tensor.get_dtype(), input_tensor.get_layout(), tile);
+    }
```

Comment:
why not?
```suggestion
        const auto tile = input_tensor.get_tensor_spec().tile();
        return Tensor(updated_storage, TensorSpec(new_shape, input_tensor.get_tensor_spec()));
    }
```

Reply:
Because we don't want tile() to be requested for tensors that don't have a tile layout. The default should be row major